datafusion/expr/src/built_in_function.rs

```diff
@@ -1550,6 +1551,46 @@ impl FromStr for BuiltinScalarFunction {
     }
 }
 
+/// `ScalarFunctionDef` is the new interface for builtin scalar functions
+/// This is an adapter between the old and new interface, to use the new interface
+/// for internal execution. Functions are planned to move into new interface gradually
+/// The function body (`execute()` in `ScalarFunctionDef`) now are all defined in
+/// `physical-expr` crate, so the new interface implementation are defined separately
+/// in `BuiltinScalarFunctionWrapper`
+impl ScalarFunctionDef for BuiltinScalarFunction {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn name(&self) -> &[&str] {
+        aliases(self)
+    }
+
+    fn input_type(&self) -> TypeSignature {
+        self.signature().type_signature
+    }
+
+    fn return_type(&self) -> FunctionReturnType {
+        let self_cloned = *self;
+        let return_type_resolver = move |args: &[DataType]| -> Result<Arc<DataType>> {
+            let result = BuiltinScalarFunction::return_type(self_cloned, args)?;
+            Ok(Arc::new(result))
+        };
+
+        FunctionReturnType::LambdaReturnType(Arc::new(return_type_resolver))
+    }
+
+    fn volatility(&self) -> Volatility {
+        self.volatility()
+    }
+
+    fn monotonicity(&self) -> Option<FuncMonotonicity> {
+        self.monotonicity()
+    }
+
+    // execution functions are defined in `BuiltinScalarFunctionWrapper`
```

Comment:
All execution code for `BuiltinScalarFunction` are in `phys-expr` crate (which depends on this crate), so they're defined elsewhere